[SR-2860] Change the way comments are exported to Doxygen

danielmartin opened the pull request:
**What's in this PR?**

SR-2860 describes a problem where some comments in Swift may generate an invalid Objective C header file `<ProjectName>-Swift.h`, rendering the project not compilable.

This PR fixes that problem by exporting comments with `///` instead of `/** */`. Swift to Objective C conversion routines in PrintAsObjC appear to print documentation comments using the Doxygen converter (`DoxygenConverter` class).

**How is it done?**

For easy reviewing, commit by commit, this PR does:
1. Small refactoring that sorts test cases from `test/Inputs/comment_to_something_conversion.swift` alphabetically, as encouraged by its header comment.
2. Add a failing test case for this bug to the suite.
3. Modify the code in `DoxygenConverter` and existing tests so that comments are exported with `///` to Doxygen.

I can squash them into a single commit after review, if necessary.

Resolves [SR-2860](https://bugs.swift.org/browse/SR-2860).

danielmartin commented:
I've tested this using a custom toolchain and some sample Xcode project which the official toolchain cannot compile. Everything works, but I'd like another pair of eyes from the people that know that particular part of the code better.

rintaro commented on test/PrintAsObjC/Inputs/comments-expected-output.h:
```diff
@@ -4,441 +4,451 @@ SWIFT_CLASS("_TtC8comments4A000")
-  Aaa.  A010.  Bbb.
-*/
+///
+///  Aaa.  A010.  Bbb.
```
Could you remove 1 space after `///`?
```
/// Aaa.  A010.  Bbb.
instead of 
///  Aaa.  A010.  Bbb.
```

rintaro replied:
Also I don't think leading/trailing standalone `///` is needed.
```objc
/// Aaa.  A010.  Bbb.
SWIFT_CLASS("_TtC8comments21A010_AttachToEntities")
```

bitjammer commented:
Along with @rintaro's comments, looks reasonable to me.

danielmartin commented:
I have removed the starting comment lines and reduced the separation between `///` and text to 1 space (HTML nested structures inside comments still get a 2 spaces indentation).

You can review it more easily by checking the diff of the unit test.

CodaFi commented:
@swift-ci please smoke test.

jrose-apple commented:
Ah, @rintaro, @bitjammer, is this good to go?

@swift-ci Please smoke test

bitjammer commented:
LGTM.

danielmartin commented:
I rebased and fixed a small conflict with a recent change in master that added `SWIFT_WARN_UNUSED_RESULT` to the export header.

Should be good to go (BTW, I can't trigger CI builds/tests from Github comments, right?).

jrose-apple commented:
@swift-ci Please smoke test and merge

jrose-apple commented:
CI testing goes with commit access, and of course "commit access" doesn't mean "commit everything without reviews".

…hm, that should probably be on [the website](https://swift.org/contributing/#commit-access).

rintaro commented on lib/IDE/CommentConversion.cpp:
```diff
@@ -538,17 +544,25 @@ class DoxygenConverter : public MarkupASTVisitor<DoxygenConverter> {
-    indent();
+    printOpeningComment();
+    // Add one space after starting comment
+    indent(1);
```
Instead of `indent(1)`/`dedent()` in constructor/destructor, can't we always `+1` in `printIndent`?
```cpp
  void printIndent() {
    for (unsigned i = 0; i < Indent + 1; ++i) {
```

rintaro commented on lib/IDE/CommentConversion.cpp:
```diff
@@ -538,17 +544,25 @@ class DoxygenConverter : public MarkupASTVisitor<DoxygenConverter> {
   DoxygenConverter(llvm::raw_ostream &OS, unsigned Indent = 0)
-    : OS(OS), Indent(Indent), IsFreshLine(true) {
+    : OS(OS), Indent(Indent), IsFreshLine(true), IsEmptyComment(true) {
     printIndent();
```
I know the initial `Indent` is always `0` in the current usage.
But `DoxygenConverter(OS, 2)` would print something like:
```objc
  ///   Aaa.
///   Bbb.
```
I think, it should be
```objc
  /// Aaa.
  /// Bbb.
```

(or, can we just remove `Indent` from the constructor paramters?)

danielmartin replied:
Now that you mention it, I don't know if the `Indent` parameter is important. I don't see any construction of `DoxygenConverter` in the codebase that uses a custom `Indent` argument.

jrose-apple commented:
This didn't actually merge, huh. Retesting just to make sure nothing broke.

@swift-ci Please smoke test

jrose-apple commented:
If you still want to make more changes, that's fine.

jrose-apple commented:
@swift-ci Please smoke test

danielmartin commented:
@jrose-apple I've removed the `Indent` parameter from the constructor, as mentioned by @rintaro, and also fixed some conflicts.

danielmartin commented:
And finally I've removed the `indent(1)/dedent()` dance to make code clearer and simpler, per @rintaro suggestions. Anything else before merging?

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Merged since this PR have been approved by David and Jordan.
Thank you Daniel for the fix and your patience!
